File: Codes from the AI-assistants/regex-redux/RegRex-ds-p4-tbr.cpp
#include <iostream>
#include <string>
#include <regex>
#include <vector>

int main() {
    // Read all input from stdin
    std::string data;
    std::string line;
    while (std::getline(std::cin, line)) {
        data += line + "\n";
    }
    
    // Record original sequence length
    int initial_length = data.length();
    
    // 1. Remove FASTA sequence descriptions and all linefeed characters
    std::regex fasta_regex(R"(>.*\n|\n)");
    std::string cleaned = std::regex_replace(data, fasta_regex, "");
    int cleaned_length = cleaned.length();
    
    // Convert to lowercase for pattern matching
    std::string sequence_lower = cleaned;
    std::transform(sequence_lower.begin(), sequence_lower.end(), 
                   sequence_lower.begin(), ::tolower);
    
    // 2. Patterns to count
    std::vector<std::string> patterns = {
        R"(agggtaaa|tttaccct)",
        R"([cgt]gggtaaa|tttaccc[acg])",
        R"(a[act]ggtaaa|tttacc[agt]t)",
        R"(ag[act]gtaaa|tttac[agt]ct)",
        R"(agg[act]taaa|ttta[agt]cct)",
        R"(aggg[acg]aaa|ttt[cgt]ccct)",
        R"(agggt[cgt]aa|tt[acg]accct)",
        R"(agggta[cgt]a|t[acg]taccct)",
        R"(agggtaa[cgt]|[acg]ttaccct)"
    };
    
    // Count matches for each pattern
    for (const auto& pattern_str : patterns) {
        std::regex pattern(pattern_str);
        auto words_begin = std::sregex_iterator(
            sequence_lower.begin(), sequence_lower.end(), pattern);
        auto words_end = std::sregex_iterator();
        int count = std::distance(words_begin, words_end);
        std::cout << pattern_str << " " << count << std::endl;
    }
    
    // 3. Magic replacements
    std::vector<std::pair<std::string, std::string>> magic_replacements = {
        {R"(tHa[Nt])", "<4>"},
        {R"(aND|caN|Ha[DS]|WaS)", "<3>"},
        {R"(a[NSt]|BY)", "<2>"},
        {R"(<[^>]*>)", "|"},
        {R"(\\|[^|][^|]*\\|)", "-"}
    };
    
    // Apply replacements sequentially
    std::string replaced_data = data;
    for (const auto& [pattern_str, replacement] : magic_replacements) {
        std::regex pattern(pattern_str);
        replaced_data = std::regex_replace(replaced_data, pattern, replacement);
    }
    
    // Remove FASTA headers and newlines from replaced data
    std::string final_cleaned = std::regex_replace(replaced_data, fasta_regex, "");
    int final_length = final_cleaned.length();
    
    // 4. Output the three lengths
    std::cout << "\n" << initial_length << std::endl;
    std::cout << cleaned_length << std::endl;
    std::cout << final_length << std::endl;
    
    return 0;
}

File: Codes from the AI-assistants/regex-redux/RegRex-ds-p4-tbr.py
import sys
import re
from collections import defaultdict

def main():
    # Read all input from stdin
    data = sys.stdin.read()
    
    # Record original sequence length
    initial_length = len(data)
    
    # 1. Remove FASTA sequence descriptions and all linefeed characters
    # Using the same simple regex pattern match-replace
    # Pattern: remove lines starting with '>' and all newlines
    cleaned = re.sub(r'>.*\n|\n', '', data)
    cleaned_length = len(cleaned)
    
    # Store for later use
    sequence_lower = cleaned.lower()
    
    # 2. Count matches for each pattern (case-insensitive)
    patterns = [
        r"agggtaaa|tttaccct",
        r"[cgt]gggtaaa|tttaccc[acg]",
        r"a[act]ggtaaa|tttacc[agt]t",
        r"ag[act]gtaaa|tttac[agt]ct",
        r"agg[act]taaa|ttta[agt]cct",
        r"aggg[acg]aaa|ttt[cgt]ccct",
        r"agggt[cgt]aa|tt[acg]accct",
        r"agggta[cgt]a|t[acg]taccct",
        r"agggtaa[cgt]|[acg]ttaccct"
    ]
    
    # Count matches for each pattern
    counts = []
    for pattern in patterns:
        # Count all matches (non-overlapping)
        count = len(re.findall(pattern, sequence_lower))
        counts.append(count)
        print(f"{pattern} {count}")
    
    # 3. Perform the magic replacements on the original data
    # Note: We use the original data (with headers and newlines) for replacements
    magic_patterns = [
        r"tHa[Nt]",
        r"aND|caN|Ha[DS]|WaS",
        r"a[NSt]|BY",
        r"<[^>]*>",
        r"\\|[^|][^|]*\\|"
    ]
    
    magic_replacements = [
        "<4>",
        "<3>",
        "<2>",
        "|",
        "-"
    ]
    
    # Apply replacements sequentially on the original data
    replaced_data = data
    for pattern, replacement in zip(magic_patterns, magic_replacements):
        replaced_data = re.sub(pattern, replacement, replaced_data)
    
    # Remove FASTA headers and newlines from the replaced data
    # (same operation as before)
    final_cleaned = re.sub(r'>.*\n|\n', '', replaced_data)
    final_length = len(final_cleaned)
    
    # 4. Output the three recorded lengths
    print(f"\n{initial_length}")
    print(cleaned_length)
    print(final_length)

if __name__ == "__main__":
    main()